design/design-doc-composition-functions-extra-resources.md

```diff
@@ -0,0 +1,1196 @@
+**Q5.1**: **WHEN** should the policy be evaluated?
+- **Option 1 (Q5.1)**: Before gathering the requested extra resources.
+- **Option 2 (Q5.1)**: After having gathered the requested extra resources.
+
+**Option 1 (Q5.1)** would be more efficient, as we would not request any
+resource in case of an invalid request. However, it would be less flexible and
+more coarse-grained, as the policy would only have to take into account the
+information provided in the request, so as per the decision taken above for
+**Q4**, only names and labels. While with **Option 2 (Q5.1)** the policy would
+have the whole resources available and could therefore be much more expressive.
+
+We decided to only consider **Option 1 (Q5.1)**, as we deemed unnecessary to
+have the extra flexibility of **Option 2 (Q5.1)**, and we could always add that
+later, if needed, but there would still be value in having a simpler API to
+cover the majority of the use cases.
+
+So, decided that the policy should be evaluated before actually gathering the
+requested extra we can answer **Q5.2**: What should be the policy's input?
+- **Option 1 (Q5.2)**: The request only.
+- **Option 2 (Q5.2)**: Also the observed state.
+
+**Option 1 (Q5.2)** would require a less expressive policy language that could
+exactly mirror the extra resources requests language, e.g. only names and
+labels, and therefore would be extremely simple to validate against the request.
+
+Note that while Functions could still implement label selectors based on the
+current observed state, e.g. a label having a key or value based on a field in
+the Composite Resource, a policy language having as input only the request
+without any other input would have to be coarser-grained.
+
+**Option 2 (Q5.2)** would allow for a much more expressive policy language, as
+it would have access to the current observed state, and as long as it "rendered"
+to something easily comparable to the requests language, it would be easy to
+implement too, e.g. JSONPath expressions, Crossplane-flavored fieldpaths or
+regexes.
+
+Kubernetes users are already used to RBAC policies being evaluated only against
+the request. So, we deemed sufficient, at least as a first implementation, to
+choose **Option 1 (Q5.2)**. We can always revisit this decision at a later stage
+if needed, but a coarser-grained and simpler policy language would still be
+useful.
+
+Note: we could reconsider **Option 2 (Q5.1)** at a later stage and add a more
+expressive policy language to be evaluated after having gathered the requested
+extra resources, for example using a full-fledged policy language such as
+[rego](https://www.openpolicyagent.org/docs/latest/policy-language/), or a
+similar syntax as the one used by [Kyverno](https://kyverno.io/), or a list of
+[Common Expression Language](https://github.com/google/cel-go) (CEL)
+expressions, which would allow the Composition author to specify much more
+complex policies.
```

Comment:
If we make such a decision, wouldn't it be unnecessary to build what is proposed in this design? If we can express exactly which resources could be requested as extra with an API in composition, why should we introduce another way to define what to request (i.e. function responding with extra_resources). 

I believe, this should be something we should consider sooner than later before committing to this design.

```yaml
- step: get-extra-resources
  functionRef:
    name: function-extra-resources
  extraResources:
      # all Something.example.test.com/v1 resources
    - apiVersion: piextensions.crossplane.io/v1alpha1
      kind: EnvironmentConfig
      # Some Kyverno like API ...
      name: "{{observed.composite.spec.environment}}"
```

FWIW, @negz already has some thoughts regarding JMESPath [here](https://github.com/crossplane/crossplane/issues/3765#issuecomment-1428878515).

Reply:
With an API as that one, it would be really hard to define iterative requests for resources, e.g. resource X named A references resource Y named B and I want resource B's content. That would require a much more complex API. While a more complex policy language evaluated against the observed state AND previously requested resources, could allow to define a coarser grained policy such as "This step can request resources of type Y only if referenced by a resource of type X ...".